nixos/modules/services/system/nscd.nix

```diff
@@ -67,6 +55,9 @@ in
         serviceConfig =
           { ExecStart = "@${pkgs.glibc.bin}/sbin/nscd nscd";
             Type = "forking";
+            User = "nscd";
+            DynamicUser = true;
```

Comment:
So that means that systemd will no longer try to use nscd to check the dynamic user? How do they implement that?

Reply:
In theory, `DynamicUser` should only work when `nscd` is on, as `nscd` dispatches to the correct `nss_systemd.so` .  So this indeed sounds like something that is not gonna work, given that `nscd` is not running when `nscd` is starting up :P

However, note that `nscd` bypasses delegating to `nscd` when doing any `nss` related syscalls itself (for good reason; it is the module implementing the `nss` syscalls in the first place), so maybe  all the `nss` related syscalls in the `nscd` binary hit the same codepath as requests from others programs that go through the `nscd` socket? Because in that case, this will work.

Non of this is documented so we could find out by reading the C source code... Or just disable `DynamicUser` for `nscd` so that we're on the safe side.

Reply:
glibc always has to be able to answer passwd queries and such whether nscd is running or not. How does it know? It tries opening the socket that nscd listens on, and if that fails, it does the lookup itself. (There's also a glibc-internal function which nscd itself calls to prevent even checking for the socket, as you noted.)

So this works fine: systemd does the lookup of a user and group named `nscd` before it has started the service, the socket doesn't work, so glibc falls back to having `libnss_files.so` handle the query in-process. If you want to configure a static user for nscd, you can do it in `/etc/passwd` and systemd will find it.

Also: any code in the `systemd` package doesn't need nscd to find its own nsswitch modules, because they're already in its library path. So without nscd, systemd would only get the wrong answers for these passwd/group lookups on systems configured with nsswitch modules that are neither the glibc builtins nor systemd's extras.

That means the real constraint here is that you can't override this `DynamicUser` setting by defining an `nscd` user in LDAP, Google OS Login, etc. Maybe that's worth documenting, but it's really a nonsense thing to even try doing...

This does make me want to try getting the username `nscd` on some large LDAP-based network somewhere and see how much I can break though. :smiling_imp:

Reply:
Yes.  Any service started under `DynamicUser` must have either `libnss_systemd.so` in its `LD_LIBRARY_PATH` or have access to the `nscd` socket in order for `DynamicUser` to work.

My worry was because `nscd` itself doesn't access the socket it wouldn't work. But obviously `nscd` has `libnss_systemd.so` in `LD_LIBRARY_PATH` so it will indeed work.  
So indeed, it seems we can leave `DynamicUser = true` with no issues.


On a related note, I think you can drop the `User = nscd`  option by the way.    Because the service name is `nscd.service` and `DynamicUser = true`, systemd already automatically sets `User = nscd`
>  The user and group name to use may be configured via
> User= and Group= (see above). If these options are not used and
> dynamic user/group allocation is enabled for a unit, the name of
> the dynamic user/group is implicitly derived from the unit name. If
> the unit name without the type suffix qualifies as valid user name
>   it is used directly, otherwise a name incorporating a hash of it is
           used.

Reply:
Good point about removing `User`, thanks! I've retested with that change and indeed the selected username is still "nscd", as expected.